### ISSUE TITLE
List of issues with OMPI master+ULT+PartCom

### ISSUE THREAD

**Issue by janciesko** [open-mpi/ompi#10459]
This tracks current issues for ULT- and PartCom support in OMPI.

ULT Support:
- [ ] Compilation fails on current master @ HEAD when configured to --with-threads={qthreads,argobots} with
` ../../opal/mca/threads/qthreads/threads_qthreads.h:29:10: fatal error: qthread.h: No such file or directory
   29 | #include "qthread.h"` Configure seems to be not setting -I/PATH_TO_ULT_LIB
- [ ] OMPI hangs when configured to --with-threads=argobots and argobots uses multiple streams and --map-by ppr:1:node thus using ucx and libevent. Requires https://github.com/shintaro-iwasaki/libevent/commits/2.0.22-abt in libevent. We need to merge that branch into libevent soon. Check why Qthreads works.
- [ ]  MPI_Init_thread hangs on current master @ HEAD when configured to --with-threads={qthreads,argobots}. Maybe this is an issue in instance/instance.h after merging sessions-related code changes.
- [ ] UCX fails asserts on POW9 and ARM: Discussed this with Howard. An assert is failing in src/ucp/core/ucp_worker.c. I'll add more info once the latest OMPI version compiles.

Partitioned Communication:
- [ ] Non equal numbers of send- and receive partitions leads to deadlock. Looking at a backtrace, it seems that the receiving rank waits for the last receive partition. MPI Partix is a reproducer when setting `DEFAULT_RECV_SEND_PARTITION_RATIO=1` to another value such as 2. 

General improvements:
- [ ] Add ULT coverage to CI testing 

Reproducers of all above:
```
git clone https://github.com/sandialabs/MPI-Partix.git
export PATH=$OMPI_INSTALL_PATH/bin
cmake ..  -DCMAKE_CXX_COMPILER=mpicxx -DQthreads_ROOT=$QTHREADS_INSTALL_PATH -DPartix_ENABLE_QTHREADS=ON
mpirun -np 2 --map-by ppr:1:node ./bench1
```

**Comment by jsquyres** [comment]
@shintaro-iwasaki Can you have a look at this?

**Comment by shintaro-iwasaki** [comment]
As I left Argonne a year ago, I might not be the right person for this, but please let me know anything I can help. Jan should have known it.

(Currently, Yanfei Guo is the POC on the Argobots side.)

**Comment by jsquyres** [comment]
@shintaro-iwasaki Gotcha.  What's Yanfei Guo's github ID?

**Comment by shintaro-iwasaki** [comment]
@yfguo He is leading the MPICH project.

**Comment by jsquyres** [comment]
@yfguo It looks like there are some errors with the qthreads implementation in Open MPI.  @shintaro-iwasaki used to be the maintainer of this, but now -- perhaps by default? -- the ownership has apparently transferred to you...?  We're having some compile issues as listed in this issue.  Can you investigate?

**Comment by shintaro-iwasaki** [comment]
Sorry for interruption. I believe @janciesko owns the Qthreads/Open MPI part.

https://github.com/open-mpi/ompi/pull/8479
https://github.com/open-mpi/ompi/pull/8500

I think Jan posted this issue to fix it later.

**Comment by jsquyres** [comment]
Ah, ok.  @janciesko Can you also update the `owner.txt` files so that we don't ping the wrong people in the future?  😄

**Comment by awlauria** [comment]
@janciesko is there any update on this? We may look to disable these for the v5.0.0 release if not completed in time, since they are new features it would not be a regression. That said, we would love to have them. We're targeting an end of summer release, so end of August/September timeframe.  Thanks!

**Comment by janciesko** [comment]
@awlauria, thanks for clarifying the time scope for this. I'll look into this and report back.

**Comment by janciesko** [comment]
@bwbarrett, can you please take a look why ompi fails to compile when enabling the use of Qtheads or Argobots? 
The following commit introduced the error: https://github.com/open-mpi/ompi/commit/6767309af2c893ff2dfb22d901f4e2fb3b7a8f87.  
Configure fails to set `OPAL_QTHREADS_INCLUDE_PATH` in `./opal/mca/threads/qthreads/threads_qthreads.h.in`
Same behavior applies to Argobots.

**Comment by hppritcha** [comment]
I am working on a fix for the first item.

**Comment by janciesko** [comment]
This https://github.com/open-mpi/ompi/blob/96fadd9d6860bd1dc89f15e88a472f310ff13c89/ompi/instance/instance.c#L226 causes a deadlock as that lock is already taken by the executing thread. The lock is previously set here: https://github.com/open-mpi/ompi/blob/96fadd9d6860bd1dc89f15e88a472f310ff13c89/ompi/instance/instance.c#L800 
https://github.com/open-mpi/ompi/pull/10720 fixes item three.

**Comment by janciesko** [comment]
Item three needs support for recursive locks in Qthreads. This is in progress.

**Comment by janciesko** [comment]
We have added a draft PR for supporting recursive locks in Qthreads. https://github.com/Qthreads/qthreads/pull/111 
With this feature, we resolve the deadlock reported in item three.

**Comment by janciesko** [comment]
We have added a PR for supporting recursive locks in Qthreads. https://github.com/Qthreads/qthreads/pull/112

**Comment by jsquyres** [comment]
Does Open MPI need to use some minimum version of Qthreads in order to ensure correctness?  If so, it might be worthwhile to enforce that in the qthreads component's `configure.m4`.

**Comment by janciesko** [comment]
@jsquyres, yes and yes, we'd need a version check for qthreads to be >=1.18.

**Comment by janciesko** [comment]
[#108032](https://github.com/open-mpi/ompi/pull/10832) fixes item three.

**Comment by janciesko** [comment]
I am marking issue 4 as complete as I cannot reproduce that issue anymore.

**Comment by janciesko** [comment]
I am marking issue 2 as complete as I cannot reproduce that issue anymore. In this case I am using the default downstream dep on libevent 2.1.12.

**Comment by janciesko** [comment]
@jsquyres, addressing the last checkbox, what methodology would you propose to test OMPI and the threading MCA when configuring OMPI with a supported ULT lib? ASFAIK we do not test this. We could add a test similar to [this one](https://github.com/sandialabs/MPI-Partix/blob/main/apps/ultcorrectness1.cpp) and test with a release version of both ULT libs?
